httprequesthandler_test.go

```diff
@@ -0,0 +1,185 @@
+		FunctionProcess: "cat",
+		ExecTimeout:     time.Duration(time.Second * 10),
+		UpstreamURL:     upstreamServer.URL,
+	}
+
+	recordHTTPRequest(t, wc, http.MethodGet, bytes.NewBuffer(expectedBody), nil)
+
+	if !bytes.Equal(actualBody, expectedBody) {
+		t.Errorf("upstream received incorrect body - want: %s, got: %s", expectedBody, actualBody)
+	}
+}
+
+func TestHTTPRequestHandler_WatchdogPassesHeaderToUpsteam(t *testing.T) {
+	expectedContentType := "text/plain"
+	actualContentType := ""
+	upstreamServer := httptest.NewServer(
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			actualContentType = r.Header.Get("Content-Type")
+		}))
+
+	wc := config.WatchdogConfig{
+		FunctionProcess: "cat",
+		ExecTimeout:     time.Duration(time.Second * 10),
+		UpstreamURL:     upstreamServer.URL,
+	}
+
+	recordHTTPRequest(t, wc, http.MethodGet, nil, map[string]string{
+		"Content-Type": expectedContentType,
+	})
+
+	if actualContentType != expectedContentType {
+		t.Errorf("upstream received incorrect content-type - want: %s, got: %s", expectedContentType, actualContentType)
+	}
+}
+
+func TestHTTPRequestHandler_WatchdogReceivesStatusFromUpstream(t *testing.T) {
+	expectedStatus := http.StatusOK
+	upstreamServer := httptest.NewServer(
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.WriteHeader(expectedStatus)
+		}))
+
+	wc := config.WatchdogConfig{
+		FunctionProcess: "cat",
+		ExecTimeout:     time.Duration(time.Second * 10),
+		UpstreamURL:     upstreamServer.URL,
+	}
+
+	rr := recordHTTPRequest(t, wc, http.MethodGet, nil, nil)
+
+	if status := rr.Code; status != expectedStatus {
+		t.Errorf("handler returned wrong status code -  want: %v, got: %v", expectedStatus, status)
+	}
+}
+
+func TestHTTPRequestHandler_WatchdogReceivesBodyFromUpstream(t *testing.T) {
+	expectedBody := []byte("openfaas rocks")
+	upstreamServer := httptest.NewServer(
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.Write(expectedBody)
+			w.WriteHeader(http.StatusOK)
+		}))
+
+	wc := config.WatchdogConfig{
+		FunctionProcess: "cat",
+		ExecTimeout:     time.Duration(time.Second * 10),
+		UpstreamURL:     upstreamServer.URL,
+	}
+
+	rr := recordHTTPRequest(t, wc, http.MethodGet, nil, nil)
+
+	if rr.Body == nil {
+		t.Errorf("handler retured nil body: expected %s", string(expectedBody))
+	}
+
+	bodyBytes, bodyErr := ioutil.ReadAll(rr.Body)
+	if bodyErr != nil {
+		t.Fatal("unable to read data from body")
+	}
+
+	if !bytes.Equal(bodyBytes, expectedBody) {
+		t.Errorf("handler returned incorrect body - want: %s, got: %s", expectedBody, bodyBytes)
+	}
+}
+
+func TestHTTPRequestHandler_StatusGatewayTimeout_WhenExecTimeouts(t *testing.T) {
+
+	upstreamServer := httptest.NewServer(
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			time.Sleep(time.Second * 2)
+		}))
+	defer upstreamServer.Close()
+
+	wc := config.WatchdogConfig{
+		FunctionProcess: "cat",
+		ExecTimeout:     time.Duration(time.Second),
```

Comment:
What is the shortest value you can use to produce the result?

Reply:
The test can still works with the timeout at 1 millisecond. This PR was updated to reflect this change.